sw/device/lib/runtime/pmp.c

```diff
@@ -0,0 +1,539 @@
+// Copyright lowRISC contributors.
+// Licensed under the Apache License, Version 2.0, see LICENSE for details.
+// SPDX-License-Identifier: Apache-2.0
+
+#include "sw/device/lib/runtime/pmp.h"
+
+#include <stdbool.h>
+#include <stddef.h>
+#include <stdint.h>
+
+#include "sw/device/lib/base/bitfield.h"
+#include "sw/device/lib/base/stdasm.h"
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6.1 Physical Memory Protection CSRs",
+// "Figure 3.28: PMP configuration register format".
+#define PMP_CFG_CSR_R 0
+#define PMP_CFG_CSR_W 1
+#define PMP_CFG_CSR_X 2
+#define PMP_CFG_CSR_A 3
+#define PMP_CFG_CSR_L 7
+
+#define PMP_CFG_FIELDS_PER_REG 4
+#define PMP_CFG_FIELD_WIDTH 8
+#define PMP_CFG_FIELD_MASK 0xffff
+
+// "Volume II: RISC-V Privileged Architectures V20190608-Priv-MSU-Ratified",
+// "3.6 Physical Memory Protection", "Address Matching".
+#define PMP_CFG_CSR_MODE_OFF 0
+#define PMP_CFG_CSR_MODE_TOR 1
+#define PMP_CFG_CSR_MODE_NA4 2
+#define PMP_CFG_CSR_MODE_NAPOT 3
+#define PMP_CFG_CSR_MODE_MASK 0x3
+
+// TODO - explanation to why macro magic is necessary (csr must be const expr).
+#define PMP_CSR_WRITE(csr, var) asm volatile("csrw " #csr ", %0;" : : "r"(var))
```

Comment:
Do we want something like this in `stdasm.h`, where we can have a single place to explain this macro magic?

Reply:
Possibly, we could pull out these and put into the `stdasm`. Another option would be to have a separate library with wrapped asm for functions like:
`mepc_get`
`mepc_set`
`mtvec_set`
`mtvec_get`

Subject to a different PR (either of those options).

Reply:
Let's do a separate PR; personally, I favor the macro version since it shouldn't be too hard to mock.